### PR TITLE
Implement the latest version of the `defineKind` API

### LINKED ISSUE #4905: new defineKind API with { state, facets } argument
## What is the Problem Being Solved?

To prevent the GC sensitivity of #4892, in today's kernel meeting we came up with a new API for `defineKind`. The current API looks like:
```js
function init(args) {
  const initialState = {}; // use 'args'
  return initialState;
}
function actualize(state) {
  const oneFacet = { meth1(args) { .. }, meth2(args) { .. } };
  // or
  const facets = {
     facetA: { meth1(args) { .. }, meth2(args) { .. } },
     facetB: { meth1(args) { .. }, meth2(args) { .. } },
  };
  return oneFacet; // or facets
}
function finish(state, cohort) {
  const { facetA } = cohort;
  // do some final wiring
}
const makeFoo = defineKind('foo iface', init, actualize, finish);
```

In the new API, the third argument to `defineKind` will be a record of unbound functions, each of which takes an initial argument `{ state, self }` or `{ state, facets }`:
```js
// one facet:
const behavior = {
  meth1: ({ state, self }, ...args) => { .. },
  meth2: ({ state, self }, ...args) => { .. },
};
// multiple facets
const behavior = {
  facetA: {
    meth1: ({ state, facets }, ...args) => { .. },
    meth2: ({ state, facets }) => { .. },
  },
  facetB: {
    meth1: ({ state, facets }) => { .. },
    meth2: ({ state, facets }, ...args) => { .. },
  },
};
```

In addition, the fourth argument will be an options bag, with the only currently-accepted option being `finish`.

```js
const makeFoo = defineKind('foo iface', init, behavior, { finish });
```

The `behavior` record is examined during `defineKind`, and a copy is made of the functions it provides (to ensure that userspace does not get control later, during actualization). Later, when `makeFoo(args)` is invoked and we need to make a new virtual object cohort, the sequence is:
* call `init(args)`, get back the initial state record
* create a `state` object with getters and setters for each property in the initial state record
* create a "context" object with initial contents `{ state, facets: {} }` (or just `{ state }` for single-facet kinds)
* for each facet in `behavior` (or just the one facet):
  * create a new object (this will become the facet)
  * for each function in the behavior record:
    * create a new function (with `Function.prototype.bind`) that curries the context object as the first argument
    * add that function as a property of the new facet object
  * add the facet to `context.facets` (or assign the one facet as `context.self`)
* build the cohort object, harden everything
* call `finish` if defined
* return the cohort to the user (return value of `makeFoo()`)

The goal is to prevent userspace from getting control during actualization, giving it a way to create per-instance objects that could be used to sense GC. The bound methods (and `state`, and the context object, and the facet objects within) are all per-instance, but they are all visible to the Kind machinery during actualization, so we can establish WeakMap links from them to the cohort record, which means they'll have the same lifetime, which prevents their use as GC sensors. By having userspace provide a table of functions at `makeKind` time, rather than executing a userspace `actualize` function at actualize time, we don't give them any opportunity to make new per-instance objects (or to even sense when actualization is taking place).

I think this will let us safely remove the `proForma` "call `actualize` anyways and throw out the result" code in the VOM, which would be lovely.

One small surprise for Kind authors is that their methods will have a different signature than the code which invokes them. The first argument is provided by the Kind machinery.

Kinds which simply want to refer to their state should get it by destructuring the first argument:
```js
const behavior = {
  incr: ({ state }, delta) => { state.count += delta; }
};
// counter.incr(delta)
```

When a kind needs to refer to other facets, it should grab that from `facets`:
```js
const behavior = {
  depositFacet: {
    deposit: ({ facets }, payment) => facets.purse.deposit(payment),
  },
  purse: {
    withdraw: ({ state }, amount) => { state.balance -= amount; ... },
    deposit: ({ state }, payment) => { ... },
    getDepositFacet: ({ facets }) => facets.depositFacet,
  },
};
```

If a single-facet virtual object needs to refer to itself, it should use `self`:
```js
const behavior = {
  subscribe: ({ state, self }) => state.publisher.subscribe(self),
};
```

## Rejected Variants

We considered `method: ({ state, ...facets }) => ...`, which would shrink some things like `getDepositFacet: ({ depositFacet }) => depositFacet` by the length of `facets.`. This would preclude the use of `state` as a facet name, and would complicate the addition of new context properties in the future. We decided that referencing other facets was infrequent enough that it wasn't worth the preclusion.

There may be value to internally representing all Kinds as multi-faceted, even those with a single facet. Independent of that, we considered having `makeKind` always be multi-facet, so authors of single-facet Kinds would need to pick a facet name. This would remove `self` from the context argument (it would always have `facets`, even if there was only one property on it). The Kind defintion would then look like:
```js
const behavior = {
  dummyName: {
    method1: ({ state, facets }) => { stuff(state, facets.dummyFacet ...} },
  },
};
```

and userspace would look like:
```js
const foo = makeFoo(args).dummyName;
```

either of which were sufficient to disqualify the idea. Instead, `defineKind` will look at the shape of `behavior` to decide whether it creates single-facet objects or multi-facet cohorts. The internals might or might not manage both in the same way, but the `behavior` record and the return value of `makeFoo` will be different for the two cases.

We considered using JavaScript's special `this` variable to hold the `{ state, facets }` context object. This would be closer to the usual prototype-based inheritance (possibly identical), but would require `behavior` be defined with concise method syntax, precluding the use of `=>` arrow functions (which have other benefits). We hope that "Far Classes" will be implemented eventually, which should avoid the surprising extra context argument and also be stored more efficiently.

We decided that `finish` is used infrequently enough that a reader of a `defineKind` definition would benefit from seeing its name appear in the options bag, and that this benefit (plus the general benefit of options bags if/when new arguments are added) justified changing the fourth `defineKind` argument from an optional `finish` function to an optional `{ finish }` bag.

## Security Considerations

Userspace must not get control during actualization, even if the behavior record they provide is a `Proxy` or has getters. We must enumerate the contents once, during `defineKind`, and assert that we get plain `Function` objects from it. We should store those Functions in the Kind record for use during actualization.

We must double-check that the act of binding a Function is not visible to that Function, perhaps through some `Proxy` trickery that I'm not aware of (does `.toString` get called for any reason?).

## Test Plan

Unit tests, ideally ones that don't pull in all of swingset, but maybe that's the easiest way to do it.

### PR DESCRIPTION
**Pull request by FUDCo**
This implements the newest version of the `defineKind` API, as laid out in #4905.

Closes #4905

### CODE REVIEW

**Comment by FUDCo** [comment]
Note: I've marked this as DRAFT because it will not pass CI, because there are a couple of uses of `defineKind` elsewhere in the source tree that need to be updated to use the new API, and what those uses are doing is complicated and weird and hard to understand.  MarkM and I will pair on fixing these on Thursday afternoon, but I wanted to get this PR up so the main substance can be reviewed.

**Comment by warner** [comment]
Does this also add the crosslinks and `unweakables`?

**Comment by FUDCo** [comment]
Yes.  Two separate commits, one for the API overhaul and one for the anti-GC-observability stuff.

**Review by dckc — COMMENTED** [review]
I defer to other reviewers.

File: packages/ERTP/src/purse.js
Comment (turadg) [inline packages/ERTP/src/purse.js:34]: the type of `purse` is no longer `Purse`, which is the use API.

we might want a helper like `WithContext` so that we can still define an external API and then use it here to constrain the implementation. e.g. `/** @type {WithContext<Purse>} */`. Though I'm sure there's a better name.

File: packages/SwingSet/docs/virtual-objects.md
Comment (warner) [inline packages/SwingSet/docs/virtual-objects.md]: Should we maybe give it the same `context` object as the `behavior` methods get? I don't think we explored this when talking about `finish`, but now I'm wondering if it would be more uniform. Not a big deal either way, for sure.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js]: `copyMethods` is given the user's `behavior` object and *returns* a `behaviorTemplate`, right? (or at least one facet's portion of a `behaviorTemplate`). Can we rename the argument and `obj` to reflect that? We need to defend against the user's `behavior` object doing something tricky (like changing its entries between two calls to `Object.entries`), but I think it can't do that because we only run `Object.entries()` against it once (in `copyMethods`), but these argument names make me think otherwise.

Also, (and maybe this is an @erights question) should we consider `harden`ing the user's `behavior` object when it is first submitted to `defineKind`? We're no longer using their records to build our facets (we build entirely new records), so having them pre-`harden`ed shouldn't interfere with the new objects being made `Far`.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js:465]: I'll definitely defer to @erights here to weigh in on the correctness. I think I assumed that there was a `Reflect` method that curried `this` and the first N arguments into a new function that could be called with `(...args)`. This looks like we're calling `Reflect.apply` every time someone invokes the method.

But now that I skim through MDN on `Reflect.*`, it seems like this is the only tool it offers, in which case this seems correct to me.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js]: I'd suggest adding `context` to the unweakables down where it's created, rather than here (multiple times).

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js]: This enables Symbol-named methods, yeah? I think we want that, at least in the long run, and anyways `Far` should be the judge of what keys are allowed, so I think `Object.entries` is exactly right.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js]: Is there any benefit to using `Object.getPrototypeOf` instead of `__proto__`? What happens if someone defines a method named `__proto__` in their `behavior`?

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js:670]: think we should `harden(toHold)` for completeness' sake?

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js]: Maybe we should have a helper function `protect(obj, target)`, which walks the prototype chain, adding each prototype to `unweakable` and establishing the `linkToCohort`, until it reaches `Object.prototype`. I had a sudden fear that someone might change `Far` down the road to have *two* hoisted prototypes, and we'd never notice the security problem.

Comment (warner) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: Also it gets called like 4 times and it's currently 4-6 lines each, so it might be more concise too.

File: packages/SwingSet/test/virtualObjects/test-representatives.js
Comment (warner) [inline packages/SwingSet/test/virtualObjects/test-representatives.js:431]: nice! thanks for testing this so thoroughly

**Comment by warner** [comment]
The `xsnap-native` change looks like a merge problem? I'm guessing it should not be in this branch.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: I wish. Two factors combine to make this not work well:
   * `Object.entries` omits non-enumerable properties
   * Object literals with symbol-named properties initialize those properties to be non-enumerable.

```js
> x = {[Symbol.asyncIterator]: () => null}
{ [Symbol(Symbol.asyncIterator)]: [Function: [Symbol.asyncIterator]] }
> Object.entries(x)
[]
> Object.getOwnPropertyDescriptors(x)
{
  [Symbol(Symbol.asyncIterator)]: {
    value: [Function: [Symbol.asyncIterator]],
    writable: true,
    enumerable: true,
    configurable: true
  }
}
```

Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: No, forget what I just said. The session I captured above shows that my second bullet is false. The new property is enumerable. But `Object.entries` omitted it anyway. Which corrected my dim memory: `Object.entries` only includes string-named properties.

In any case, I agree that `Far` should know what it accepts and others should make minimal assumptions about what that is. Use `Reflect.ownKeys` or `Object.getOwnPropertyDescriptors` in order to minimize such assumptions.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: Oh god yes! Please avoid `__proto__`. Not only can `Object.prototype.__proto__` be overridden, Caridy convinced us and tc39 that secure environments should probably remove it on initialization. SES does not yet do so, but probably will and should (attn @kriskowal @mhofman ). tc39 made it normative optional, so implementations are allowed to not provide it in the first place (as a future SES/XS may elect to do).

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: I expect `FarClasses` will need something like this anyway.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: > should we consider `harden`ing the user's `behavior` object when it is first submitted to `defineKind`? We're no longer using their records to build our facets (we build entirely new records), so having them pre-`harden`ed shouldn't interfere with the new objects being made `Far`.

That's a good point. Yes, we should preemptively `harden`.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js:465]: We created our own `uncurryThis`, but it just calls `Reflect.apply` to solve the underlying problem.

File: packages/SwingSet/docs/virtual-objects.md
Comment (FUDCo) [reply packages/SwingSet/docs/virtual-objects.md]: The same thought occurred to me as I was writing it, but ultimately I decided that the gratuitous destructuring didn't really pay for itself.  However, that's not a strongly held belief on my part and I'm happy to change it to work the other way if that's the consensus.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: I'm plucking the methods out of the `behavior` object that's passed in to guard against other crap that the user might have hung off the object.  After extracting the methods there's really no benefit to `harden`ing (or otherwise fiddling with) the original object.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js:465]: This is the incantation that @erights gave me.  I have no investment in it.  If a better recipe comes along, I'm happy to use it.  In the meantime I think this is the best we can do.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: This was a case of my own ignorance coming into play.  I had never before, in years of JS development, ever had occasion to want to get to the prototype of an object.  Fortunately, this is a very easy fix.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js:670]: In practical terms that would result in neither harm nor benefit, but I like the aesthetic of positivism w.r.t immutability.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: So are we going to explicitly allow symbol-named methods here and leave it to `Far` to reject them?
What about symbol-named facets?

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: I started with something like that and it was a fail, because it turns out that things that need to make `unweakable` and things that need `linkToCohort` are not as correlated as one might imagine.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: Thinking about this further, symbol-named facets aren't straightforward, because I need to put them in order and sorting Symbols is a pain.  In any event, I don't think there's a use case for this aside from obsessive generality, so let's take the idea of symbol-named facets off the table for now.

Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: On deeper analysis, symbol-named methods are kind of PITA too, though if we think it's worth the effort I can do it.  But I really don't think it's worth the effort.

Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: Definitely no symbol-named facets. The record naming facets is a copyRecord, and the properties of a copyRecord must be string-named own enumerable properties, so they're exempt from all these painful differences between ways to enumerate own properties. So this restriction is architectural, not expedient.

> symbol-named methods here and leave it to `Far` to reject them?

`Far` does not reject symbol-named methods. Nor does it reject methods named by non-enumerable properties.

But we currently have only one case we need, `@@asyncIterator`, which we could continue to special case for now. If postponing general support for symbol-named methods gets us to MN1 faster, I'm all for postponing. But I think our distributed computational model should eventually support symbol-named methods in general.

File: packages/SwingSet/docs/virtual-objects.md
Comment (erights) [reply packages/SwingSet/docs/virtual-objects.md]: My inclination is the other way, to use the same context object. Not because the destructuring is compelling, but so there's one less thing to learn.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (erights) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: I agree there is no strong reason to `harden` this one. As long as the extraction is careful not to be confusable by a self-mutating behavior object. For example, doing the extraction all at once with an `Object.getOwnPropertyDescriptors` and then fully ignoring the original object would be safe.

Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: That's an interesting wrinkle I hadn't thought about: is `Object.getOwnPropertyDescriptors` atomic in a way that iterating over the array returned by `Object.entries` is not?  Thinking about it, that seems right, since if you use `Object.entries`, each pass of the iteration gives the object being iterated over the opportunity to do something funky mid-iteration if one or more of the properties is a getter.  (Of course in order to use the value returned by `Object.getOwnPropertyDescriptors` you have to iterate over _it_ using `Object.entries` or one of its brethren, which is kind of a pain and feels at first glance like you're just recursing on the problem -- but presumably you are now in exclusive possession of _that_ object in a way that you are not in exclusive possession of the original object).  On the other hand, it's unclear to me what mischief somebody could cause by having the behavior object modify itself during iteration, since `defineKind` is only going to iterate over the behavior object once, so any modification is as if that was what it was supposed to do all along.  All `defineKind` cares about is obtaining a set of method functions, so if the object squirmed around somehow while being iterated I don't see why `defineKind` would care. It got its bag of methods from the thing its caller gave it, so where's the harm? Seems to me the worst case is they wrote some weird code that they might not understand later, but anyone can confuse themselves if they work at it.

File: packages/SwingSet/docs/virtual-objects.md
Comment (FUDCo) [reply packages/SwingSet/docs/virtual-objects.md]: OK, done.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: If I can put in symbol-named methods in less than, say, an hour's work I'll do it.  I suspect I probably can.

Comment (FUDCo) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js]: A weird thought occurred to me as I was pondering implementation: as long as we do not support symbol-named methods in our over-the-wire message encoding, someone can use symbol-named methods as a kind of private-to-the-local-vat scoping technique.  I'm not saying this is a good idea, but I can imagine people doing it.  Then, when we turn that feature on, we break their imagined security perimeter.  (Even once we support symbols in the method selector portion of a message, I presume they would still be able to use non-registered, non-well-known symbols for this purpose.)  This makes me wonder if we should hold off implementing symbol-named methods until we are also implementing symbols as method selectors in messages.

**Comment by FUDCo** [comment]
I think this is now ready for final review.

Note to @michaelfig @Chris-Hibbert @turadg: would you please take a look at `packages/run-protocol/src/vaultFactory/vault.js`?  I had to make fairly invasive changes to update it for the new `defineKind` API,, which I don't believe changed any of the semantics (it passes all the tests, anyway), but given the radical transformation I prefer somebody take a look who actually understands the substance of the underlying code.

**Comment by turadg** [comment]
> would you please take a look at `packages/run-protocol/src/vaultFactory/vault.js`

Thanks for the ping. I don't see any change in semantics, which is good. I see a separation of `selfBehavior` and `helperBehavior`, which is good. I do have some concerns.

1) Git blame churn. Many lines in this file are being touched. I suggest we have a codeowner of run-protocol be the author of the commit for `vault.js` so that git blame tracks back. I'm happy to take that on.

2) Loss of JSDoc. JSDoc (at least in my IDE) is on the value of the object and when a function is composed into a new object (e.g. `selfBehavior`) it doesn't come along. I think each self behavior function should be nested in the `selfBehavior` object. Without that the function documentation will be lost in the calling context.

3) Untyped `facets`. That can be fixed with,
```js
/**
 * @typedef {{
 *   state: ImmutableState & MutableState,
 *   facets: {
 *     self: import('@agoric/vat-data/src/types').FunctionsMinusContext<typeof selfBehavior>,
 *     helper: import('@agoric/vat-data/src/types').FunctionsMinusContext<typeof helperBehavior>,
 *   },
 * }} MethodContext
 */
```

4) Loss of argument names in types. With the type transformation to remove the first parameter, the names of the other parameters become `args_0` etc. I don't know what to do about this. Maybe there's a trick to do in `FunctionsMinusContext`. I'll research this.

5) Loss of type information within behavior functions. Previously the type of `state` was declared once and shared in each behavior function because the value was shared. Now the context passed in can be _anything_ unless there's an annotation constraining it. I don't know how to solve this without annotating each function.

6) Loss of calback types like `OfferHandler`. This could be solved with a `PlusContext` partner for `MinusContext`. E.g. `{WithContext<OfferHandler>}` below.
https://github.com/Agoric/agoric-sdk/blob/df924b3fbdf1e63b586dba4aaec2f3b988588b44/packages/run-protocol/src/vaultFactory/vault.js#L360

Not all the above has to be solved in this PR. My blocking requests are to:
- inline the function definitions to the behavior objects
- type `facets` in MethodContext

File: packages/ERTP/src/payment.js
Comment (turadg) [inline packages/ERTP/src/payment.js]: mind filing a ticket for this?

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vault.js:112]:
```suggestion
 *   facets: {
 *     self: import('@agoric/vat-data/src/types').FunctionsMinusContext<typeof selfBehavior>,
 *     helper: import('@agoric/vat-data/src/types').FunctionsMinusContext<typeof helperBehavior>,
 *   },
```

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vault.js]: please move the behavior functions within the object that will hold them

**Review by turadg — CHANGES_REQUESTED** [review]

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vault.js]: Needs a return type to prevent circular inference.
```suggestion
  * @returns {string} success message
  */
```

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js]: Can we rename `behaviorTemplate` here to `behavior`, to match the caller's signature? It's confusing to me that this claims to *accept* `behaviorTemplate` rather than returning it.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/liveslots/virtualObjectManager.js:646]: because I just wondered about it again, and it took me a minute to convince myself it was ok, let's add a one-line comment here
```
# 'context' does not need a linkToCohort because it's held by the bound methods

**Review by warner — CHANGES_REQUESTED** [review]
Two small changes to make, plus the `xsnap-native` submodule change needs to not be in this PR. Rebasing onto trunk might clear that out, but if not grab me and we can figure it out.

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (FUDCo) [reply packages/run-protocol/src/vaultFactory/vault.js]: Turns out tsc wanted `Promise<string>` rather than `string`.  I _think_ this is right...

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (FUDCo) [reply packages/run-protocol/src/vaultFactory/vault.js]: I'm unclear what you mean by "the behavior functions".   Did you mean inline _all_ the behavior functions or just the two that you highlighted in this comment?  Because they weren't all inlined before, by any means.

File: packages/ERTP/src/payment.js
Comment (FUDCo) [reply packages/ERTP/src/payment.js]: Done.  #5018

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (FUDCo) [reply packages/run-protocol/src/vaultFactory/vault.js:112]: I thought this would be a straightforward edit, but it turned out to yield a bunch of follow-on type errors that I have not the first clue how to resolve.  I'm going to put it back the way it was and ask you to toss it in with all the other TS backfilling you've already flagged here.  Possibly also this could be added to the work in #5018?

**Comment by FUDCo** [comment]
How do I expunge the `xsnap-native` nonsense that somehow got sucked into this PR?  Rebasing didn't seem to do the trick.

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/vault.js:112]: Ok

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/vault.js]: I did mean all of them. They weren't all inlined in object creation but they were all nested and that's a property that I want to maintain. They also all had JSDoc before and the new object construction breaks that unless they're inlined.

Comment (FUDCo) [reply packages/run-protocol/src/vaultFactory/vault.js]: Done!  Please give it a look.

File: packages/SwingSet/src/liveslots/virtualObjectManager.js
Comment (warner) [reply packages/SwingSet/src/liveslots/virtualObjectManager.js:646]: oh, crud, I got the comment backwards, and it's going to confuse me in its current state

please replace with:

`'context' does not need a linkToCohort because it holds the facets (which hold the cohort)`

sorry

**Review by warner — APPROVED** [review]
r+ module that one extra comment change

**Comment by FUDCo** [comment]
> r+ module that one extra comment change

OK, but I think your prior attempt at the comment was also correct.

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/vault.js]: Perfect, thank you.

**Review by turadg — APPROVED** [review]

**Pull request merged**
